### ISSUE TITLE
Impossible to use S3 compatibles in version 4.15, works in 4.13

### ISSUE THREAD

**Issue by Tyki** [strapi/strapi#19085]
### Required System information

- Node.js version: 20.5
- NPM version: 9.8.0
- Strapi version: 4.15
- Database: Mysql
- Operating system: Ubuntu 22.04.1
- Is your project Javascript or Typescript: TS

### Describe the bug

I have the following error when I try to upload a file to S3 compatible (Scaleway) on version 4.15

### Steps to reproduce the behavior

1. Install latest version of aws s3 plugin (4.15)
2. Configure S3 compatible (documentation configuration, nothing more)

```
upload: {
      config: {
        provider: "aws-s3",
        providerOptions: {
          accessKeyId: env("SCALEWAY_ACCESS_KEY_ID"),
          secretAccessKey: env("SCALEWAY_ACCESS_SECRET"),
          endpoint: env("SCALEWAY_ENDPOINT"), // e.g. "s3.fr-par.scw.cloud"
          params: {
            Bucket: env("SCALEWAY_BUCKET")
          }
        }
      }
    }
```
```
SCALEWAY_ACCESS_KEY_ID=XXXXXXXXXXXXXXXXx
SCALEWAY_ACCESS_SECRET=xxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxxxxxx
SCALEWAY_ENDPOINT="s3.fr-par.scw.cloud"
SCALEWAY_BUCKET="xx-bucket"
```

3. Try to upload an image
4. Here is the following error : 

```
strapi_1                 | [2023-12-19 09:22:47.883] http: GET /upload/folders?sort=createdAt:DESC&page=1&pageSize=10&pagination[pageSize]=-1&filters[$and][0][parent][id][$null]=true (39 ms) 200
strapi_1                 | node:internal/errors:496
strapi_1                 |     ErrorCaptureStackTrace(err);
strapi_1                 |     ^
strapi_1                 | 
strapi_1                 | TypeError [ERR_INVALID_URL]: Invalid URL
strapi_1                 |     at new NodeError (node:internal/errors:405:5)
strapi_1                 |     at new URL (node:internal/url:778:13)
strapi_1                 |     at parseUrl (/usr/app/node_modules/@smithy/url-parser/dist-cjs/index.js:7:38)
strapi_1                 |     at toEndpointV1 (/usr/app/node_modules/@smithy/middleware-endpoint/dist-cjs/adaptors/toEndpointV1.js:12:38)
strapi_1                 |     at Object.customEndpointProvider [as endpoint] (/usr/app/node_modules/@smithy/middleware-endpoint/dist-cjs/resolveEndpointConfig.js:10:99)
strapi_1                 |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
strapi_1                 |     at async Promise.all (index 1)
strapi_1                 |     at async Upload.__uploadUsingPut (/usr/app/node_modules/@aws-sdk/lib-storage/dist-cjs/Upload.js:66:26)
strapi_1                 |     at async Upload.__doConcurrentUpload (/usr/app/node_modules/@aws-sdk/lib-storage/dist-cjs/Upload.js:128:28)
strapi_1                 |     at async Promise.all (index 0) {
strapi_1                 |   input: 's3.fr-par.scw.cloud',
strapi_1                 |   code: 'ERR_INVALID_URL'
strapi_1                 | }
```

### Expected behavior

A success upload of the image on the bucket

### Additional context
It works on 4.13 perfectly

**Comment by Th1nhNg0** [comment]
Same issue

**Comment by dimitrisganotis** [comment]
In the new version you need an endpoint that starts with https://.
So try https://s3.fr-par.scw.cloud

**Comment by dawidchlodnicki** [comment]
Same issue, set "https://" requires region

**Comment by christiancp100** [comment]
What we did in the AWS provider was update the AWS dependency, going from AWS V2 to AWS V3. This update of the version changed the format of the URLs, as you said and it's specified in [this document](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#virtual-hosted-style-access). 
TLDR: AWS recommends the new URL format (virtual-hosted–style URI with the format ) since the Path-style URI will be deprecated at some point. [This is another post](https://aws.amazon.com/es/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/) that justifies better their decision. If Scaleway doesn't support using the new format and you want to make it work with 4.15.x versions without migrating anything, you can set a `baseUrl` property in the plugin with your desired URL format, as shown in the example below:
```
  upload: {
    config: {
      provider: 'aws-s3',
      providerOptions: {
        accessKeyId: env('AWS_ACCESS_KEY_ID'),
        secretAccessKey: env('AWS_ACCESS_SECRET'),
        region: env('AWS_REGION'),
        baseUrl: 'https://s3.{region}.amazonaws.com/{bucket}, // THIS LINE IS THE IMPORTANT ONE
        params: {
          ACL: env('AWS_ACL', 'public-read'),
          signedUrlExpires: env('AWS_SIGNED_URL_EXPIRES', 15 * 60),
          Bucket: env('AWS_BUCKET'),
        },
      },
      actionOptions: {
        upload: {},
        uploadStream: {},
        delete: {},
      },
    },
  },
```

However, if you can do as @dimitrisganotis says and use the https format specifying the Scaleway region you're using, that would be a better option. 

If you need anything else let us know and sorry for the late reply 🙏🏼

**Comment by RebootGG** [comment]
Hey @christiancp100, even when using the provided example, I have exactly the same problem, using scaleway S3 buckets.


```
        s3Options: {
          accessKeyId: env("SCALEWAY_ACCESS_KEY_ID"),
          secretAccessKey: env("SCALEWAY_ACCESS_SECRET"),
          baseUrl: 'https://s3.{region}.scw.cloud/{bucket}',
          region: env("SCALEWAY_REGION"),
          params: {
            ACL: 'public-read',
            signedUrlExpires: '15 * 60',
            Bucket: env("SCALEWAY_BUCKET"),
          },
        },
```

```
SCALEWAY_REGION=fr-par
SCALEWAY_BUCKET=my-bucket 
```

When trying to upload a file: I have the following error : 

```
Error: getaddrinfo ENOTFOUND my-bucket.s3.fr-par.amazonaws.com
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:107:26)
    at GetAddrInfoReqWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
  errno: -3008,
  code: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'my-bucket.s3.fr-par.amazonaws.com',
  '$metadata': { attempts: 1, totalRetryDelay: 0 }
}
```


Seems to me that the `amazonaws.com` in the hostname maybe comes from the amazon SDK ?

**Comment by RebootGG** [comment]
Of course ! On my Scaleway dashboard, the bucket endpoint is like so : 

`https://my-bucket.s3.fr-par.scw.cloud`

So it would be `https://<bucket-name>.s3.<region>.scw.cloud`

To be honest, I think it would be simpler to be able to pass directly the whole URL, like I was trying to do in the `baseUrl` property.

That means that my config could only look like this : 

```
        s3Options: {
          credentials: {
            accessKeyId: env("SCALEWAY_ACCESS_KEY_ID"),
            secretAccessKey: env("SCALEWAY_ACCESS_SECRET"),
          },
          baseUrl: 'https://my-bucket.s3.fr-par.scw.cloud',
          params: {
            ACL: 'public-read',
          },
        },
```

**Comment by exneval** [comment]
just want to confirm 4.14.3 still works, upgrade to 4.15.x, 4.16.x, 4.17.x, no longer works

**Comment by gmoocto** [comment]
We cannot get it to work with our S3 compatible service either : 

Setting a `baseUrl` gives the following error : 
"InvalidAccessKeyId: The AWS Access Key Id you provided does not exist in our records."

The error does not seem to originate from our S3 compatible service (on which I can log S3 interaction)
So I assume it's trying to reach AWS (on which the access key would, indeed, be invalid)

**Comment by CEI-Devs** [comment]
Don't work with the 4.19.1 too

Any news concerning the provider Scaleway ?

**Comment by gmoocto** [comment]
Hello ! Any news regarding this issue? 
We are currently holding back on updates, we're stuck on 4.14

**Comment by BaptisteKaliop** [comment]
We are stuck in 4.13 also because of this issue

**Comment by christiancp100** [comment]
> Hello ! Any news regarding this issue? We are currently holding back on updates, we're stuck on 4.14

Hey! I'm probably going to work on this during next week, hopefully get to a solution. Thank you and sorry for the waiting 🙏🏼

**Comment by christiancp100** [comment]
Hello again! I've been trying to configure the provider with Scaleway and it is working on my side, I attach a video and the configuration that I use.

```
// plugins.js
  upload: {
    config: {
      sizeLimit: 25 * 1024 * 1024 * 1024,
      provider: 'aws-s3',
      providerOptions: {
        s3Options: {
          accessKeyId: env('ACCESS_KEY'),
          secretAccessKey: env('SECRET_KEY'),
          region: env('REGION'),
          endpoint: env('UPLOADS_ENDPOINT'),
          params: {
            ACL: env('AWS_ACL', 'public-read'),
            signedUrlExpires: env('AWS_SIGNED_URL_EXPIRES', 15 * 60),
            Bucket: env('AWS_BUCKET', 'strapi-test'),
          },
          s3ForcePathStyle: true,
        },
      },
      actionOptions: {
        upload: {},
        uploadStream: {},
        delete: {},
      },
    },
  },
```

```
# .env
ACCESS_KEY=<scaleway-access-key>
SECRET_KEY=<scaleway-secret-key>
UPLOADS_ENDPOINT=https://s3.fr-par.scw.cloud # CHECK HERE THE HTTPS:// IN FRONT OF THE URL AND NO BUCKET NAME ON IT
BUCKET_NAME=strapi-test
```

This is the video of the demo: https://www.loom.com/share/347de035e19140e28d2759669929e1ef?sid=80d1d2b8-27bf-4b01-ac6c-bebbdc30faed
![Screenshot 2024-03-01 at 15 52 44](https://github.com/strapi/strapi/assets/16720378/7f1ecdf3-0b38-41b2-868e-f4d1a84d2f1c)

**Comment by zifeo** [comment]
`s3ForcePathStyle` has been renamed to `forcePathStyle` in v3 of the s3 sdk. The following works with a local minio:
```typescript
export default ({ env }) => {
  return {
    upload: {
      config: {
        provider: "aws-s3",
        providerOptions: {
          s3Options: {
            accessKeyId: env("AWS_ACCESS_KEY_ID"),
            secretAccessKey: env("AWS_ACCESS_SECRET"),
            region: env("AWS_REGION"),
            endpoint: env("AWS_ENDPOINT"),
            forcePathStyle: true,
            params: {
              ACL: "private",
              signedUrlExpires: 15 * 60,
              Bucket: env("AWS_BUCKET"),
            },
          },
        },
        actionOptions: {
          upload: {},
          uploadStream: {},
          delete: {},
        },
      },
    },
  };
};
```

**Comment by christiancp100** [comment]
> `s3ForcePathStyle` has been renamed to `forcePathStyle` in v3 of the s3 sdk. The following works with a local minio:
> 
> ```ts
> export default ({ env }) => {
>   return {
>     upload: {
>       config: {
>         provider: "aws-s3",
>         providerOptions: {
>           s3Options: {
>             accessKeyId: env("AWS_ACCESS_KEY_ID"),
>             secretAccessKey: env("AWS_ACCESS_SECRET"),
>             region: env("AWS_REGION"),
>             endpoint: env("AWS_ENDPOINT"),
>             forcePathStyle: true,
>             params: {
>               ACL: "private",
>               signedUrlExpires: 15 * 60,
>               Bucket: env("AWS_BUCKET"),
>             },
>           },
>         },
>         actionOptions: {
>           upload: {},
>           uploadStream: {},
>           delete: {},
>         },
>       },
>     },
>   };
> };
> ```

Yep, you're right there, I'm going to update the documentation for this. Although for this issue it doesn't really matter, since it can work with and without `forcePathStyle`. Thank you very much for this!

**Comment by christiancp100** [comment]
I've put together a Strapi project in the 4.20.3 version (current latest) with a working example of a Scaleway configuration to upload assets to a Scaleway bucket. Try it out and let me know if you encounter any errors on your side, otherwise I will close this issue in the following days: 
[s3-test.zip](https://github.com/strapi/strapi/files/14493192/s3-test.zip)

**Comment by BaptisteKaliop** [comment]
Looks like there is still issues from @RebootGG above with 4.20

I'll try to find some times this week to test on my side since it was not planned to upgrade strapi anymore due to this issue

**Comment by christiancp100** [comment]
> Looks like there is still issues from @RebootGG above with 4.20
> 
> I'll try to find some times this week to test on my side since it was not planned to upgrade strapi anymore due to this issue

Yes, but @RebootGG configuration is not accurate, he's putting the region in the endpoint url, and that's not right.

**Comment by RebootGG** [comment]
I followed your example [above](https://github.com/strapi/strapi/issues/19085#issuecomment-1973338899) @christiancp100 :

 From your config : 
> `endpoint: env('UPLOADS_ENDPOINT'),`

You added the region as well in your .env variable :
> `UPLOADS_ENDPOINT=https://s3.fr-par.scw.cloud # CHECK HERE THE HTTPS:// IN FRONT OF THE URL AND NO BUCKET NAME ON IT`


I only changed the names of my variables, not the values you indicated.

In the .zip you shared above, there is also the region in the`.env.example`

`UPLOADS_ENDPOINT=https://s3.<your-region-here>.scw.cloud` 

---------

Does that mean the url for the `endpoint` has to be formatted like this ? 

```
   endpoint: 'https://s3.scw.cloud',
```

(Just tried and got another error which is better than a plain 500)

`Error [ERR_TLS_CERT_ALTNAME_INVALID]: Hostname/IP does not match certificate's altnames: Host: scw.cloud. is not in the cert's altnames: DNS:ensemble-bbb.scaleway.com`

**Comment by christiancp100** [comment]
> I followed your example [above](https://github.com/strapi/strapi/issues/19085#issuecomment-1973338899) @christiancp100 :
> 
> From your config :
> 
> > `endpoint: env('UPLOADS_ENDPOINT'),`
> 
> You added the region as well in your .env variable :
> 
> > `UPLOADS_ENDPOINT=https://s3.fr-par.scw.cloud # CHECK HERE THE HTTPS:// IN FRONT OF THE URL AND NO BUCKET NAME ON IT`
> 
> I only changed the names of my variables, not the values you indicated.
> 
> In the .zip you shared above, there is also the region in the`.env.example`
> 
> `UPLOADS_ENDPOINT=https://s3.<your-region-here>.scw.cloud`
> 
> Does that mean the url for the `endpoint` has to be formatted like this ? (without the region)
> 
> ```
>    endpoint: 'https://s3.scw.cloud',
> ```
> 
> (Just tried and got another error which is better than a plain 500)
> 
> `Error [ERR_TLS_CERT_ALTNAME_INVALID]: Hostname/IP does not match certificate's altnames: Host: scw.cloud. is not in the cert's altnames: DNS:ensemble-bbb.scaleway.com`

My endpoint is this: https://s3.fr-par.scw.cloud

**Comment by RebootGG** [comment]
So is mine [here](https://github.com/strapi/strapi/issues/19085#issuecomment-1973398733)

> `SCALEWAY_ENDPOINT=https://s3.fr-par.scw.cloud`

Is there a diff I'm not seeing ? 😅

**Comment by Th1nhNg0** [comment]
> `s3ForcePathStyle` has been renamed to `forcePathStyle` in v3 of the s3 sdk. The following works with a local minio:
> 
> ```ts
> export default ({ env }) => {
>   return {
>     upload: {
>       config: {
>         provider: "aws-s3",
>         providerOptions: {
>           s3Options: {
>             accessKeyId: env("AWS_ACCESS_KEY_ID"),
>             secretAccessKey: env("AWS_ACCESS_SECRET"),
>             region: env("AWS_REGION"),
>             endpoint: env("AWS_ENDPOINT"),
>             forcePathStyle: true,
>             params: {
>               ACL: "private",
>               signedUrlExpires: 15 * 60,
>               Bucket: env("AWS_BUCKET"),
>             },
>           },
>         },
>         actionOptions: {
>           upload: {},
>           uploadStream: {},
>           delete: {},
>         },
>       },
>     },
>   };
> };
> ```

This really fix my problem. I'm using s3 compatible storage and add the `forcePathStyle` is fix the problem. I still need to add `region` to config but any value is ok since my provider dont have the region field.

**Comment by christiancp100** [comment]
> So is mine [here](https://github.com/strapi/strapi/issues/19085#issuecomment-1973398733)
> 
> > `SCALEWAY_ENDPOINT=https://s3.fr-par.scw.cloud`
> 
> Is there a diff I'm not seeing ? 😅

You need to specify the region in the `providerOptions` AND in the endpoint, and the bucket name in the providerOptions but NOT in the endpoint url.

**Comment by RebootGG** [comment]
> > So is mine [here](https://github.com/strapi/strapi/issues/19085#issuecomment-1973398733)
> > > `SCALEWAY_ENDPOINT=https://s3.fr-par.scw.cloud`
> > 
> > 
> > Is there a diff I'm not seeing ? 😅
> 
> You need to specify the region in the `providerOptions` AND in the endpoint, and the bucket name in the providerOptions but NOT in the endpoint url.

I'm really confused, I've checked multiple times, it seems to me that's exactly what I did ? 

> You need to specify the region in the providerOptions 

✅ 
> AND in the endpoint 

✅ 
> and the bucket name in the providerOptions  

✅

> but NOT in the endpoint url. 

✅ 

```
    providerOptions: {
        s3Options: {
          accessKeyId: env("SCALEWAY_ACCESS_KEY_ID"),
          secretAccessKey: env("SCALEWAY_ACCESS_SECRET"),
          region: "fr-par",  // region "fr-par" in the providerOptions
          endpoint: "https://s3.fr-par.scw.cloud", // region in the endpoint but no bucket name
          params: {
            ACL: "public-read",
            signedUrlExpires: 15 * 60,
            Bucket: env("SCALEWAY_BUCKET"), // bucket name in the providerOptions  
          },
          forcePathStyle: true,
        },
      },
```

```
SCALEWAY_ACCESS_KEY_ID=**** # correct SCW key
SCALEWAY_ACCESS_SECRET=**** # correct SCW secret
# SCALEWAY_ENDPOINT=https://s3.fr-par.scw.cloud 
SCALEWAY_BUCKET=**** # just the bucket name, e.g. "my-bucket"
```

It seems like the problem might just be on my end, since it's working for others. 

I'll stop hijacking this topic and wait to see if other people can't make it work too.

**Comment by gmoocto** [comment]
Hello, 
We got it working on our side by updating `s3ForcePathStyle` to `forcePathStyle` and adding `https://` to our endpoint value.
It's working both locally using a minio instance and with our S3 compatible service (ECS).

Thank you for your help ! 🙏

**Comment by BaptisteKaliop** [comment]
Works for me on Strapi v4.21, thanks a lot @christiancp100 for the time took for investigation
